cirq-core/cirq/ops/gateset.py

```diff
@@ -0,0 +1,264 @@
+        )
+
+    def _value_equality_values_(self) -> Any:
+        # `isinstance` is used to ensure the a gate type and gate instance is not compared.
+        return isinstance(self.gate, raw_types.Gate), self.gate, self.name, self.description
+
+
+@value.value_equality()
+class Gateset:
+    """Gatesets represent a collection of GateFamily objects.
+
+    Gatesets are useful for
+        a) Describing the set of allowed gates in a human readable format
+        b) Validating a given gate / optree against the set of allowed gates
+
+    Gatesets rely on the underlying GateFamily's for both description and
+    validation purposes.
+
+    The default GateFamily can be used to description and validation of the
+    two most common cases:
+        a) A specific gate instance is allowed
+        b) All instances of a specific gate class are allowed.
+
+    Users can define custom gate families for more complex scenarios and
+    use the GateFamily objects as members of a Gateset.
+    """
+
+    def __init__(
+        self,
+        *gates: Union[Type[raw_types.Gate], raw_types.Gate, GateFamily],
+        name: Optional[str] = None,
+    ) -> None:
+        """Inits Gateset.
+
+        Accepts a list of gates, each of which should be either
+            a) `cirq.Gate` subclass
+            b) `cirq.Gate` instance
+            c) `cirq.GateFamily` instance
+
+        `cirq.Gate` subclasses and instances are converted to the default
+        `cirq.GateFamily(gate=g)` instance and thus a default name and
+        description is populated.
+
+        Args:
+            gates: A list of `cirq.Gate` subclasses / `cirq.Gate` instances /
+            `cirq.GateFamily` instances to initialize the Gateset.
+            name: (Optional) Name for the Gateset. Useful for description.
+        """
+        self._name = name
+        self._gates = frozenset(
+            [g if isinstance(g, GateFamily) else GateFamily(gate=g) for g in gates]
+        )
+
+    @property
+    def name(self) -> Optional[str]:
+        return self._name
+
+    @property
+    def gates(self) -> FrozenSet[GateFamily]:
+        return self._gates
+
+    def __contains__(self, item: Union[raw_types.Gate, raw_types.Operation]) -> bool:
+        """Check for containment of `item` in any of the member `GateFamily`.
+
+        This method just forwards containment checks to the underlying
+        `GateFamily` objects. It returns `True` if any of the member gate
+        families can accept `item`.
+
+        To validate whether a `cirq.CircuitOperation`/`cirq.OP_TREE`/
+        `cirq.Circuit` is made up entirely of gates from this Gateset, use the
+        `validate` method given below.
+
+        Args:
+            item: The `cirq.Gate` or `cirq.Operation` instance to check
+                containment for.
+        """
+        for gate_family in self.gates:
+            if item in gate_family:
+                return True
+        return False
```

Comment:
Can we make this O(1) instead of O(n) ?

Reply:
Since derived gate families can have custom predicates, it's not possible to make it O(1) for the general case. We could do optimizations specific to the default gate family to check in O(1) for the trivial case and then do an O(n) if no match was found, but I would rather defer such optimizations to the point where we can show that this O(n) is becoming a performance bottleneck. Note that I'd generally expect the number of gate families in a gateset to be fairly small (<= 10 on average?) so the `n` in `O(n)` is more like a small constant. 

Sure, it can lead to a `10x` performance degradation factor when validating a large circuit, but do we already have circuits large enough where this is a bottleneck?

Reply:
I wouldn't put it past some brave users to create gatesets with as many as 50 or more gates in a gateset. Do you think there might be any way we could keep it O(1) for a vanilla `GateFamily` instance and then special case any of the subclassed `GateFamily` instances ?

Reply:
Made containment O(1) for vanilla `GateFamily` instances. PTAL.